src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+                self.config.bulk_write_max_entries,
+                MONGODB_BULK_WRITE_SIZE_LIMIT
+            )));
+        }
+
+        if let Err(err) = self.config.common.collection_name.parse::<Namespace>() {
+            return Err(SinkError::Config(anyhow!(err).context(format!(
+                "invalid collection.name {}",
+                self.config.common.collection_name
+            ))));
+        }
+
+        // checking reachability
+        let client = self.config.common.build_client().await?;
+        let client = ClientGuard::new(self.param.sink_name.clone(), client);
+        client
+            .database("admin")
+            .run_command(doc! {"hello":1}, None)
+            .await
+            .map_err(|err| {
+                SinkError::Mongodb(anyhow!(err).context("failed to send hello command to mongodb"))
+            })?;
+
+        if self.config.drop_collection_name_field && self.config.collection_name_field.is_none() {
+            return Err(SinkError::Config(anyhow!(
+                    "collection.name.field must be specified when collection.name.field.drop is enabled"
+                )));
+        }
+
+        // checking dynamic collection name settings
+        if let Some(coll_field) = &self.config.collection_name_field {
+            let fields = self.schema.fields();
+
+            let coll_field_index = fields
+                .iter()
+                .enumerate()
+                .find_map(|(index, field)| {
+                    if &field.name == coll_field {
+                        Some(index)
+                    } else {
+                        None
+                    }
+                })
+                .ok_or(SinkError::Config(anyhow!(
+                    "collection.name.field {} not found",
+                    coll_field
+                )))?;
+
+            if fields[coll_field_index].data_type() != risingwave_common::types::DataType::Varchar {
+                return Err(SinkError::Config(anyhow!(
+                    "the type of collection.name.field {} must be varchar",
+                    coll_field
+                )));
+            }
+
+            if !self.is_append_only && self.pk_indices[0] == coll_field_index {
+                return Err(SinkError::Config(anyhow!(
+                    "collection.name.field {} must not be equal to the primary key field",
+                    coll_field
+                )));
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn new_log_sinker(&self, writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(MongodbSinkWriter::new(
+            format!("{}-{}", writer_param.executor_id, self.param.sink_name),
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            self.is_append_only,
+        )
+        .await?
+        .into_log_sinker(writer_param.sink_metrics))
+    }
+}
+
+pub struct MongodbSinkWriter {
```

Comment:
Can you merge `MongodbSinkWriter` and `MongodbPayloadWriter`? They seem to be similar

Reply:
Indeed the `MongodbSinkWriter` is too simple. The reason I separated them is that I don't want the buffering logic, which is related to the builder, to be handled in MongodbSinkWriter. I'd suggest to keeping them coexistent as they give us a chance to create a fresh `MongodbPayloadWriter` each time when checkpointing in the future. The current implementation does not create a fresh `MongodbPayloadWriter` each time is because I want to keep the capacity of builder's hashmap.